spyder_kernels/customize/namespace_manager.py

```diff
@@ -55,23 +55,36 @@ def __init__(self, filename, namespace=None, current_namespace=False,
         self._previous_main = None
         self._reset_main = False
         self._file_code = file_code
+        self._saved_globals = None
 
     def __enter__(self):
         """
         Prepare the namespace.
         """
         # Save previous __file__
         if self.ns_globals is None:
+            global_namespace, local_namespace = _get_globals_locals()
             if self.current_namespace:
-                self.ns_globals, self.ns_locals = _get_globals_locals()
+                self.ns_globals = global_namespace
+                self.ns_locals = local_namespace
                 if '__file__' in self.ns_globals:
                     self._previous_filename = self.ns_globals['__file__']
                 self.ns_globals['__file__'] = self.filename
             else:
                 ipython_shell = get_ipython()
+                shell_namespace = ipython_shell.user_ns
                 main_mod = ipython_shell.new_main_mod(
                     self.filename, '__main__')
-                self.ns_globals = main_mod.__dict__
+                if shell_namespace is global_namespace:
+                    # Run in the global namespace
+                    self._saved_globals = shell_namespace.copy()
+                    # Here we clear all variables, let's hope ipykernel
+                    # is not too mad
+                    shell_namespace.clear()
```

Comment:
IPython saves several hidden variables that don't come from running code. This could break that.

Reply:
Yes but there shouldn’t be any ipython code running between enter and exit. All the code runs in the same thread does it not? I agree that this is a risk

Reply:
> Yes but there shouldn’t be any ipython code running between enter and exit. All the code runs in the same thread does it not?

Yeah, I think so.

> I agree that this is a risk

Have you tested this for a while to see if it introduces problems?

Reply:
I think we need to call `get_ipython().init_user_ns` after clearing the namespace to add the hidden variables back.